cpp/src/barretenberg/dsl/acir_proofs/acir_proofs.cpp

```diff
@@ -154,4 +189,200 @@ bool verify_proof(
+ *                                    This will be empty if this is the first proof to be recursively.
+ * @param output_aggregation_obj_buf - Two G1 points that the top-level verifier needs to run a pairing upon to complete
+ * verification
+ */
+size_t verify_recursive_proof(uint8_t const* proof_buf,
+                              uint32_t proof_length,
+                              uint8_t const* vk_buf,
+                              uint32_t vk_length,
+                              uint32_t num_public_inputs,
+                              uint8_t const* input_aggregation_obj_buf,
+                              uint8_t** output_aggregation_obj_buf)
+{
+    const size_t NUM_AGGREGATION_ELEMENTS = acir_format::RecursionConstraint::NUM_AGGREGATION_ELEMENTS;
+
+    bool inner_aggregation_all_zero = true;
+    std::vector<barretenberg::fr> aggregation_input(acir_format::RecursionConstraint::AGGREGATION_OBJECT_SIZE);
+    for (size_t i = 0; i < acir_format::RecursionConstraint::AGGREGATION_OBJECT_SIZE; i++) {
+        aggregation_input[i] = barretenberg::fr::serialize_from_buffer(&input_aggregation_obj_buf[i * 32]);
+        inner_aggregation_all_zero &= (aggregation_input[i].is_zero());
+    }
+
+    acir_format::aggregation_state_ct previous_aggregation;
+    if (!inner_aggregation_all_zero) {
+        std::array<acir_format::bn254::fq_ct, NUM_AGGREGATION_ELEMENTS> aggregation_elements;
+        for (size_t i = 0; i < NUM_AGGREGATION_ELEMENTS; ++i) {
+            aggregation_elements[i] =
+                acir_format::bn254::fq_ct(acir_format::field_ct(aggregation_input[NUM_AGGREGATION_ELEMENTS * i]),
+                                          acir_format::field_ct(aggregation_input[NUM_AGGREGATION_ELEMENTS * i + 1]),
+                                          acir_format::field_ct(aggregation_input[NUM_AGGREGATION_ELEMENTS * i + 2]),
+                                          acir_format::field_ct(aggregation_input[NUM_AGGREGATION_ELEMENTS * i + 3]));
+            aggregation_elements[i].assert_is_in_field();
+        }
+        // If we have a previous aggregation object, assign it to `previous_aggregation` so that it is included
+        // in stdlib::recursion::verify_proof
+        previous_aggregation.P0 = acir_format::bn254::g1_ct(aggregation_elements[0], aggregation_elements[1]);
+        previous_aggregation.P1 = acir_format::bn254::g1_ct(aggregation_elements[2], aggregation_elements[3]);
+        previous_aggregation.has_data = true;
+    } else {
+        previous_aggregation.has_data = false;
+    }
+
+    acir_format::Composer composer;
+
+    std::vector<acir_format::field_ct> proof_fields(proof_length / 32);
+    std::vector<acir_format::field_ct> key_fields(vk_length / 32);
+    for (size_t i = 0; i < proof_length / 32; i++) {
+        // TODO(maxim): The stdlib pairing primitives fetch the context from the elements being used in the pairing
```

Comment:
Issue for this TODO can be found here: https://github.com/AztecProtocol/barretenberg/issues/447